File: SocketComm.Examples/SendArray/TwoDClient.cs
using static SocketComm.Cmd;

namespace SocketComm.Examples.SendArray;

public class TwoDClient
{
    public static void Main()
    {
        FPipe p = new FPipe("/tmp/TwoDArray.pipe");
        p.Connect();
        p.WriteCmd(Ready);
        Cmd c;
        c = p.ReadCmd();
        while (c == Receive)
        {
            int[,] i = (int[,]) p.ReceiveValue((val) =>
                SimpleConversion.BytesToArray(val, SimpleConversion.BytesToNumber<int>));
            Console.Write("Received [");
            for (int a = 0; a < i.GetLength(0); a++)
            {
                Console.Write("[");
                for (int b = 0; b < i.GetLength(1); b++)
                {
                    Console.Write(i[a,b] + ", ");
                }
                Console.Write("], ");
            }
            Console.WriteLine("]");
            p.WriteCmd(Ready);
            p.ExpectCmd(Go);
            var res = Bench(i);
            p.WriteCmd(Done);
            Console.WriteLine($"Sending {res}");
            p.SendValue(res, SimpleConversion.NumberToBytes);
            c = p.ReadCmd();
        }

        p.Dispose();
    }

    public static int Bench(int[,] i)
    {
        var sum = 0;
        foreach (var e in i)
        {
            sum += e;
        }

        return sum;
    }
}

File: SocketComm.Examples/SendArray/TwoDClient.java
import java.nio.ByteBuffer;

public class TwoDClient {
    public static void main(String[] args) throws Exception {
        FPipe p = new FPipe("/tmp/TwoDArray.pipe");
        p.WriteCmd(Cmd.Ready);
        Cmd c;
        c = p.ReadCmd();
        while (c == Cmd.Receive)
        {
            var i = (Integer[][])p.ReceiveValue((buf)-> SimpleConversion.BytesToArray(buf, ByteBuffer::getInt));
            System.out.print("Received [");
            for (var e : i) {
                System.out.print("[");
                for(var f: e){
                    System.out.print(f +", ");
                }
                System.out.print("], ");
            }
            System.out.println("]");
            p.WriteCmd(Cmd.Ready);
            p.ExpectCmd(Cmd.Go);
            var res = Bench(i);
            p.WriteCmd(Cmd.Done);
            System.out.printf("Sending %d\n", res);
            p.SendValue(res, (val)-> ByteBuffer.allocate(4).putInt(val));
            c = p.ReadCmd();
        }
        p.close();
    }

    public static int Bench(Integer[][] i){

        var sum = 0;
        for (var e : i){
            for(var f: e){
                sum += f;
            }
        }
        return sum;
    }
}
